File: examples/test2/b2clnt.java
/* $Id: b2clnt.java,v 1.5 1999/08/03 01:57:38 janssen Exp $
 BeginILUCopyright

 Copyright (c) 1991-1999 Xerox Corporation.  All Rights Reserved.

 Unlimited use, reproduction, modification, and distribution of this
 software and modified versions thereof is permitted.  Permission is
 granted to make derivative works from this software or a modified
 version thereof.  Any copy of this software, a modified version
 thereof, or a derivative work must include both the above copyright
 notice of Xerox Corporation and this paragraph.  Any distribution of
 this software, a modified version thereof, or a derivative work must
 comply with all applicable United States export control laws.  This
 software is made available AS IS, and XEROX CORPORATION DISCLAIMS ALL
 WARRANTIES, EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION THE
 IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 PURPOSE, AND NOTWITHSTANDING ANY OTHER PROVISION CONTAINED HEREIN, ANY
 LIABILITY FOR DAMAGES RESULTING FROM THE SOFTWARE OR ITS USE IS
 EXPRESSLY DISCLAIMED, WHETHER ARISING IN CONTRACT, TORT (INCLUDING
 NEGLIGENCE) OR STRICT LIABILITY, EVEN IF XEROX CORPORATION IS ADVISED
 OF THE POSSIBILITY OF SUCH DAMAGES.

 EndILUCopyright
*/
/* Chris Jacobi, May 4, 1998 5:39 pm PDT */
/* Last edited by Mike Spreitzer October 9, 1998 1:36 pm PDT */

/* THIS HAS NOT BEEN DEBUGGED */
/* THIS HAS NOT BEEN DEBUGGED */
/* THIS HAS NOT BEEN DEBUGGED */
/* THIS HAS NOT BEEN DEBUGGED */
/* THIS HAS NOT BEEN DEBUGGED */
/* THIS HAS NOT BEEN DEBUGGED */
/* THIS HAS NOT BEEN DEBUGGED */

package Test;
 
/** 
 * Main class uses same name as c program important because 
 * I have no idea what it is supposed to do and therefore the
 * association to its origin is necessary to figure out whether
 * this program does what it should...
 *
 * Don't use this as example java code.
 */ 
public class b2clnt {
    
    static void usage() {
        System.err.println(
            "Usage: java Test.b2clnt pCall bTimeout pPush len nCalls [-r] [-v]"
            );
        System.err.println("   times in milliseconds");
        System.exit(1);
    } //usage
    

    static void usage(Throwable t) {
        if (t!=null) {
            System.err.println(t);
            t.printStackTrace(System.err);
        }
        usage();
    } //usage
    
    
    public static void main(String argv[]) {
        long pCallTime = 0;
        long bTimeoutTime = 0;
        long pPushTime = 0;
        int len = 0;
        int nCalls = 1000000000;
        boolean verbose = false;
        boolean retry = false;
        System.out.println("starting b2clnt.main");
        //
        //read in command line arguments
        try {
            if (argv.length<5) {
                usage();
            }
            pCallTime = (new Long(argv[0])).intValue();
            bTimeoutTime = (new Long(argv[1])).intValue();
            pPushTime = (new Long(argv[2])).intValue();
            len = (new Integer(argv[3])).intValue();
            nCalls = (new Integer(argv[4])).intValue();
            if (nCalls<=0) {nCalls = 1000000000;}
            for (int i = 5; i<argv.length; i++) {
                if ("-v".equalsIgnoreCase(argv[i])) {
                    verbose = true;
                } else if ("-r".equalsIgnoreCase(argv[i])) {
                    retry = true;
                } else {
                    usage();
                }
            }
        } catch (NumberFormatException ne) {
            usage(ne);
        }
        System.out.println(
           "got the input: pCallTime="  + pCallTime 
           + " bTimeoutTime=" + bTimeoutTime
           + " pPushTime=" + pPushTime
           + " len=" + len
           + " nCalls=" + nCalls
           + " verbose=" + verbose
           + " retry=" + retry);
        //
        //invoke operations on proper thread
        BThread bt = new BThread(
            pCallTime, bTimeoutTime, pPushTime, len, nCalls, verbose, retry
            );
        bt.start();
        //
        //using join to be able to print the "done" message
        System.out.println("Waiting to join...");
        try {
            bt.join();
        } catch (java.lang.InterruptedException ie) {
            System.err.println("InterruptedException caught by main: " + ie);
            ie.printStackTrace(System.err);
            System.exit(1);
        }
        System.out.println("main is done");
    } //main
    
} //b2clnt


class CallBackImpl implements Batcher2.CallBack {
    
    public void Reply(long r, int x, long s) {
        System.out.println(
            "Reply(" + r  
            + ", " + x
            + ", " + s
            +")"
            );
            
    } //Reply

} //CallBackImpl


class PushThread extends Thread {
    
    BThread bt;
    
    PushThread(BThread bt) {
        this.bt = bt;
    } //constructor
    
    /** PushNow */
    public void run() {
        System.out.println("Started PushThread");
        while (! bt.done) {
            long now = System.currentTimeMillis();
            if (bt.verbose) {
                System.out.println("PushNow(" + now + ")");
            }
            bt.doPush(bt.batcher);
            try {
                Thread.sleep(bt.pPushTime);
            } catch (java.lang.InterruptedException ie) {
                System.out.println("InterruptedException " + ie);
            }
        }
    } //run
    
} //PushThread


/**
 * I can't decipher the c source whether this should be batching or not..
 */
class CallThread extends xerox.ilu.IluServiceThread {
    
    BThread bt;
    
    CallThread(BThread bt) {
        this.bt = bt;
    } //constructor
    
    
    public void run() {
        System.out.println("Started CallThread");
        xerox.ilu.IluBatcher.setCurrentBatcher(bt.batcher);
        xerox.ilu.IluPipeline.setCurrentPipeline(bt.pipeline);
        for (int i = 0;  i<bt.nCalls; i++) {
            long now = System.currentTimeMillis();
            if (bt.done) {return;}
            try {
                if (bt.verbose){System.out.println("do 1 request");}
                bt.svr.Request(now, i, bt.phil, bt.replyTo);
                if (bt.verbose){System.out.println("did 1 request");}
            } catch (xerox.ilu.IluSystemException e) {
                System.err.println("Push() => " + e);
                if (! bt.retry) {
                    e.printStackTrace(System.err);
                    System.exit(1);
                }
            }
            try {
                Thread.sleep(bt.pCallTime);
            } catch (java.lang.InterruptedException ie) {
                System.out.println("InterruptedException " + ie);
            }
        }
        bt.done = true;
    } //run
    
} //CallThread


/** Running as a thread to enable hidden call data */ 
class BThread extends xerox.ilu.IluServiceThread {

    long pCallTime;
    long bTimeoutTime;
    long pPushTime;
    int len;
    int nCalls;
    boolean verbose = false;
    boolean retry = false;
    
    public BThread(
            long pCallTime, 
            long bTimeoutTime, 
            long pPushTime,
            int len,
            int nCalls,
            boolean verbose,
            boolean retry) {
        this.pCallTime = pCallTime;
        this.bTimeoutTime = bTimeoutTime;
        this.pPushTime = pPushTime;
        this.len = len;
        this.nCalls = nCalls;
        this.verbose = verbose;
        this.retry = retry;
    }//constructor
    
    Batcher2.CallBack replyTo;
    Batcher2.Server svr; //a surrogate object
    byte[] phil;
    xerox.ilu.IluPipeline pipeline;
    xerox.ilu.IluBatcher batcher;
    boolean done = false;
    
    /** called as main */ 
    public void run() {
        System.out.println("Starting BThread");
        xerox.ilu.IluServer s = null;
        try {
            s = xerox.ilu.IluServer.createServer(null);
            System.out.println("Created IluServer");
            replyTo = new CallBackImpl();
            Batcher2.CallBackStub.registerTrueObject( 
                null, 
                replyTo, 
                s
                );
            System.out.println("registered replyTo");
            phil = new byte[this.len];
            svr = (Batcher2.Server) xerox.ilu.IluSimpleBinding.lookup(
                    "Batcher2-Server", 
                    "it", 
                    Batcher2.ServerStub.iluClass()
                    );
            System.out.println("got svr");
            batcher = xerox.ilu.IluBatcher.alloc(
                (int) bTimeoutTime, (pPushTime > 0L)
                );
            xerox.ilu.IluBatcher.setCurrentBatcher(batcher);
            System.out.println("setCurrentBatcher");
            pipeline = xerox.ilu.IluPipeline.alloc();
            xerox.ilu.IluPipeline.setCurrentPipeline(pipeline);
            System.out.println("setCurrentPipeline");
            //
            if (pPushTime > 0) {
                PushThread pt = new PushThread(this);
                pt.start();
            }
            //
            CallThread ct = new CallThread(this);
            ct.start();
            //
            xerox.ilu.IluBatcher.setCurrentBatcher(null);
            xerox.ilu.IluPipeline.setCurrentPipeline(null);
            //
            if (pPushTime > 0) {
                doPush(batcher);
            }
            //
            //wait some time
            long dt = 0;
            if (bTimeoutTime>0) {
                dt = pPushTime;
            } else if (pPushTime>0) {
                dt = bTimeoutTime;
            } else if (pPushTime < bTimeoutTime) {
                dt = pPushTime; //looks funny to me
            } else {
                dt = bTimeoutTime; //looks funny to me
            }
            dt = dt+10000;
            Thread.sleep(dt);
        } catch (java.lang.InterruptedException i) {
            System.err.println("InterruptedException: " + 1);
            i.printStackTrace(System.err);
            System.exit(1);
        } catch (xerox.ilu.IluSystemException e) {
            System.err.println("exception: " + e);
            e.printStackTrace(System.err);
            System.exit(1);
        }
    } //run
    
    
    void doPush(xerox.ilu.IluBatcher batcher) {
        try {
            batcher.pushBatcher();
        } catch (xerox.ilu.IluSystemException e) {
            System.err.println("Push() => " + e);
            if (!retry) {
               e.printStackTrace(System.err);
               System.exit(1); 
            }
        }
    } //doPush
        
} //BThread

File: examples/test2/b2clnt.c
/** $Id: b2clnt.c,v 1.3 1999/08/03 01:57:37 janssen Exp $
 BeginILUCopyright

 Copyright (c) 1991-1999 Xerox Corporation.  All Rights Reserved.

 Unlimited use, reproduction, modification, and distribution of this
 software and modified versions thereof is permitted.  Permission is
 granted to make derivative works from this software or a modified
 version thereof.  Any copy of this software, a modified version
 thereof, or a derivative work must include both the above copyright
 notice of Xerox Corporation and this paragraph.  Any distribution of
 this software, a modified version thereof, or a derivative work must
 comply with all applicable United States export control laws.  This
 software is made available AS IS, and XEROX CORPORATION DISCLAIMS ALL
 WARRANTIES, EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION THE
 IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 PURPOSE, AND NOTWITHSTANDING ANY OTHER PROVISION CONTAINED HEREIN, ANY
 LIABILITY FOR DAMAGES RESULTING FROM THE SOFTWARE OR ITS USE IS
 EXPRESSLY DISCLAIMED, WHETHER ARISING IN CONTRACT, TORT (INCLUDING
 NEGLIGENCE) OR STRICT LIABILITY, EVEN IF XEROX CORPORATION IS ADVISED
 OF THE POSSIBILITY OF SUCH DAMAGES.

 EndILUCopyright
*/
/* Last edited by Mike Spreitzer October 9, 1998 12:06 pm PDT */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>	/* for exit() */

#ifndef WIN32
#include <unistd.h>	/* for sleep() */
#endif

#include "Batcher2.h"

Batcher2_Server svr;
Batcher2_CallBack replyTo;
float           pCallF, bTimeoutF, pPushF;
ilu_FineTime    pCallT, bTimeoutT, pPushT;
long unsigned   nCalls = 1000000000;
ilu_cardinal    callx = 0;
int             verbose = 0, retry = 0;
ilu_refany      callAlarm, pushAlarm = NULL, pauseAlarm;
ILU_C_Batcher   batcher;
ILU_C_Pipeline  pl;
Batcher2_Bytes  phil = {0};
int             stop, done = 0;

static void DoPush(void)
{
  ILU_C_ENVIRONMENT env;
  ILU_C_PushBatcher(batcher, &env);
  if (!ILU_C_SUCCESSFUL(&env)) {
    fprintf(stderr, "Push() => %s\n", ILU_C_EXCEPTION_ID(&env));
    if (!retry)
      exit(1);
  }
}

static void PushNow(ilu_private rock)
{
  ilu_FineTime    now = ilu_FineTime_Now();
  ilu_FineTime    next = ilu_FineTime_Add(now, pPushT);
  if (done)
    return;
  if (verbose)
    printf("PushNow(@%lu:%lu, next=%lu:%lu)\n",
	   (long unsigned) now.ft_s,
	   (long unsigned) now.ft_t,
	   (long unsigned) next.ft_s,
	   (long unsigned) next.ft_t);
  ilu_SetAlarm(pushAlarm, next, PushNow, NULL);
  DoPush();
  return;
}

static void CallNow(ilu_private rock)
{
  ILU_C_ENVIRONMENT env;
  ilu_FineTime    now = ilu_FineTime_Now();
  ilu_FineTime    next = ilu_FineTime_Add(now, pCallT);
  ilu_longcardinal r;
  if (callx < nCalls) {
    ilu_SetAlarm(callAlarm, next, CallNow, NULL);
    ILU_LONGCARD_HIGH_WORD(&r) = now.ft_s;
    ILU_LONGCARD_LOW_WORD(&r) = now.ft_t;
    Batcher2_Server_Request(svr, r, callx, &phil, replyTo, &env);
    callx++;
    if (!ILU_C_SUCCESSFUL(&env)) {
      fprintf(stderr, "Push() => %s\n", ILU_C_EXCEPTION_ID(&env));
      if (!retry)
	exit(1);
    }
  } else {
    done = 1;
    ILU_C_StopRun(&stop);
  }
  return;
}

static void UnPause(ilu_private rock)
{
  ILU_C_StopRun(&stop);
}

void
server_Batcher2_Server_Request(Batcher2_Server self,
			       Batcher2_Time r,
			       ilu_cardinal x,
			       Batcher2_Bytes * phil,
			       Batcher2_CallBack replyTo,
			       ILU_C_ENVIRONMENT * Env)
{
  return;
}

void
server_Batcher2_CallBack_Reply(Batcher2_CallBack self,
			       Batcher2_Time r,
			       ilu_cardinal x,
			       Batcher2_Time s,
			       ILU_C_ENVIRONMENT * Env)
{
  printf("Reply(%lu:%lu, %lu -> %lu:%lu)\n",
	 (unsigned long) ILU_LONGCARD_HIGH_WORD(&r),
	 (unsigned long) ILU_LONGCARD_LOW_WORD(&r),
	 (unsigned long) x,
	 (unsigned long) ILU_LONGCARD_HIGH_WORD(&s),
	 (unsigned long) ILU_LONGCARD_LOW_WORD(&s)
  );
  return;
}

int
main(int argc, char *argv[])
{
  ILU_C_Server    s;
  ILU_C_ENVIRONMENT env;
  int             i;
  long unsigned   len;
  if (argc < 5)
    goto usage;
  if (sscanf(argv[1], "%f", &pCallF) != 1)
    goto usage;
  if (sscanf(argv[2], "%f", &bTimeoutF) != 1)
    goto usage;
  if (sscanf(argv[3], "%f", &pPushF) != 1)
    goto usage;
  if (sscanf(argv[4], "%lu", &len) != 1)
    goto usage;
  if (argc > 5 && argv[5][0] != '-') {
    if (sscanf(argv[i = 5], "%lu", &nCalls) != 1)
      goto usage;
  } else
    i = 4;
  for (i++; i < argc; i++) {
    if (strcmp(argv[i], "-v") == 0)
      verbose = 1;
    else if (strcmp(argv[i], "-r") == 0)
      retry = 1;
    else
      goto usage;
  }
  Batcher2__Initialize();
  Batcher2__InitializeServer();
  pCallT = ilu_FineTime_FromDouble(pCallF);
  bTimeoutT = ilu_FineTime_FromDouble(bTimeoutF);
  pPushT = ilu_FineTime_FromDouble(pPushF);
  Batcher2_Bytes_Init(&phil, len, NULL);
  phil._length = len;
  s = ILU_C_InitializeServer(NULL, NULL, NULL, NULL, ILU_NIL, ilu_TRUE);
  replyTo = Batcher2_CallBack__CreateTrue(NULL, s, NULL);
  callAlarm = ilu_CreateAlarm();
  pushAlarm = ilu_CreateAlarm();
  pauseAlarm = ilu_CreateAlarm();
  svr = ILU_C_LookupObject("Batcher2-Server", "it",
			   Batcher2_Server__MSType);
  if (svr == NULL) {
    fprintf(stderr, "Unable to import server!\n");
    exit(1);
  }
  batcher = ILU_C_CreateBatcher(bTimeoutT, pPushF > 0.0, &env);
  pl = ILU_C_CreatePipeline(&env);
  ILU_C_SetBatcherContext(batcher);
  ILU_C_SetPipelineContext(pl);
  if (pPushF > 0.0)
    PushNow(NULL);
  CallNow(NULL);
  ILU_C_StoppableRun(&stop);
  ILU_C_SetBatcherContext(NULL);
  ILU_C_SetPipelineContext(NULL);
  if (pPushF > 0.0)
    DoPush();
  ILU_C_ReleaseBatcher(batcher, &env);
  ILU_C_ReleasePipeline(pl, &env);
  {
    ilu_FineTime    now, dt, fini;
    now = ilu_FineTime_Now();
    dt = ((!bTimeoutF) ? pPushT :
	  (!pPushF) ? bTimeoutT :
	  (pPushF < bTimeoutF) ? pPushT : bTimeoutT);
    dt = ilu_FineTime_Add(dt, ilu_FineTime_FromDouble(10.0));
    fini = ilu_FineTime_Add(now, dt);
    ilu_SetAlarm(pauseAlarm, fini, UnPause, NULL);
  }
  ILU_C_StoppableRun(&stop);
  return 0;
usage:
  fprintf(stderr,
	  "Usage: %s pCall bTimeout pPush len [nCalls] [-r] [-v]\n",
	  argv[0]);
  return (1);
}
